### PR TITLE
Rewriting static_for, adding overload that iterates over enums

### PR DESCRIPTION
**Pull request by myrrc**
Changelog category (leave one):

- Not for changelog (changelog entry is not required.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by myrrc** [comment]
@Mergifyio update

File: base/base/static_for.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#include "FnTraits.h"
+#include "CTArray.h"
+
+namespace detail
+{
+template <class T, T Begin, T... Is>
+constexpr bool static_for(auto && f, std::integer_sequence<decltype(Begin), Is...>)
+{
+    return (
+        std::forward<decltype(f)>(f)(
```
Comment (novikd) [inline base/base/static_for.h:14]: Why do you need `std::forward` here?

File: src/Common/tests/gtest_enum_reflection.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    EXPECT_EQ(fmt::to_string(Foo::Bar), "Bar");
+    EXPECT_EQ(fmt::to_string(Foo::Baz), "Baz");
+    EXPECT_EQ(fmt::to_string(Foo::other), "other");
+    EXPECT_EQ(fmt::to_string(Foo::CAPS), "CAPS");
+    EXPECT_EQ(fmt::to_string(Foo::snake_case), "snake_case");
+
+    EXPECT_EQ(fmt::format("{} value", Foo::Bar), "Bar value");
+}
+
+template <Foo I> constexpr int foo() { return static_cast<int>(I); }
```
Comment (novikd) [inline src/Common/tests/gtest_enum_reflection.cpp:17]: Use `to_underlying`

File: base/base/constexpr_helpers.h [hunk 3]
```diff
@@ -1,32 +0,0 @@
-#pragma once
-
-#include <type_traits>
-#include <utility>
-
-template <auto Val, decltype(Val)... List>
-inline constexpr bool static_in_v = std::disjunction_v<std::bool_constant<Val == List>...>;
-
-template <typename Func, typename Arg>
-bool func_wrapper(Func && func, Arg && arg)
```
Comment (novikd) [inline base/base/constexpr_helpers.h:10]: Why did you remove `func_wrapper`?

Comment (myrrc) [reply base/base/constexpr_helpers.h:10]: Unneeded as all functors in static_for should return bool (simpler semantics).

File: src/Common/tests/gtest_enum_reflection.cpp (hunk 2 above)
Comment (myrrc) [reply src/Common/tests/gtest_enum_reflection.cpp:17]: Ok for test purposes.

File: base/base/static_for.h (hunk 1 above)
Comment (myrrc) [reply base/base/static_for.h:14]: The "always forward arguments while taking an &&" rule.

File: base/base/CTArray.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+constexpr CTArray<T, N + U> operator||(const CTArray<T, N> & arr, const CTArray<T, U> & other)
+{
+    std::array<T, N + U> st;
+    std::copy_n(arr.begin(), N, st.begin());
+    std::copy_n(other.begin(), U, st.begin() + N);
+
+    return CTArray<T, N + U>(st);
+}
+
+constexpr auto MakeCTArray(auto arg, auto ...args)
```
[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:38]: CTArray is meant to be used only as a template argument with integral or integral-like types, so that's also ok.

File: base/base/CTArray.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+
+/// Compile-time array that can be used as a template argument.
+template <class T, size_t N>
+struct CTArray
```
[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: 1. You can't specialize std::array||std::array (and even if you could, it would be quite controversial).
2. A distinct type is better for semantics.
3. calling constexpr find on CTArray (contains) is simpler than calling an explicit find on std::array

File: base/base/CTArray.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:38]: I can add a requires expression if you think it's vital.

File: base/base/constexpr_helpers.h (hunk 3 above)
Comment (novikd) [reply base/base/constexpr_helpers.h:10]: No, I think that's not OK. It leads to writing unnecessary code (for example, changes in `DictionaryStructure.h`).

File: base/base/static_for.h (hunk 1 above)
Comment (novikd) [reply base/base/static_for.h:14]: `std::forward` is supposed to be used only for passing function arguments.

File: base/base/constexpr_helpers.h (hunk 3 above)
Comment (myrrc) [reply base/base/constexpr_helpers.h:10]: The only isage is in Dictionary structure, others always return bool. And even in dictionary structure my solution (exit iteration after type discovery) is better than previous (iterate over all types even after match was found).

File: base/base/static_for.h (hunk 1 above)
Comment (myrrc) [reply base/base/static_for.h:14]: Ok, I'll remove it here.

File: base/base/CTArray.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: 2. Why?
1. and 2. Why can't you just introduce `constexpr` functions for this purposes?

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: 2. CTArray is a distinct type suitable for using in template arguments.  std::array is a general-purpose-type. In user code it's better to separate types suitable for specialized tasks,
3. calling contains() on a constexpr object is more simple than declaring a constexpr object and making a constexpr it/bool as a find result. Member functions are prefered in STL over free functions, so I'll stick to that.

**Comment by novikd** [comment]
Currently, I see no reasons for using `CTArray` instead of `std::array`.

File: src/Common/tests/gtest_enum_reflection.cpp (hunk 2 above)
Comment (novikd) [reply src/Common/tests/gtest_enum_reflection.cpp:17]: Well, yes, but it's better to use standard solutions instead of inventing your own bicycles.

**Comment by myrrc** [comment]
Another reason.: useful operator|| for concatenation. Every free function would be a more ugly solution. || is used in tests and in other PRs that depend on current, e.g. https://github.com/ClickHouse/ClickHouse/pull/29690/files#diff-327a0de53df2de18dff948ffd4b7fdae50243402505dac868c9884535fc0b12eR19

File: base/base/CTArray.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: Why do you need to use array as template argument in the first place?

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: https://github.com/ClickHouse/ClickHouse/pull/29690/files#diff-327a0de53df2de18dff948ffd4b7fdae50243402505dac868c9884535fc0b12eR43

Main use is over here --the only way is to iterate over a container supplied as a template argument if you need compiler-time dispatching.

File: base/base/constexpr_helpers.h (hunk 3 above)
Comment (novikd) [reply base/base/constexpr_helpers.h:10]: `static_for` is an utility function. It doesn't matter how is it used now. The only important thing is how it'll be used in the future after your change.
By removing `func_wrapper` you say that in the future nobody never will want to iterate over all the elements of the container.
I see that as a bad API design, because you can not guarantee that.

Comment (myrrc) [reply base/base/constexpr_helpers.h:10]: Handling void return types introduces additional complexity both to function implementation and to function semantics which won't pay off.

File: src/Common/tests/gtest_enum_reflection.cpp (hunk 2 above)
Comment (myrrc) [reply src/Common/tests/gtest_enum_reflection.cpp:17]: https://en.cppreference.com/w/cpp/utility/to_underlying Not supported in standard.

File: base/base/CTArray.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: I see situation as follows:
1. You want to process each element of the container in the compile time.
2. You need to call `static_for` on this container.
And here we've got a question. Why do you need to pass this container as a template parameter while you can pass it as an `static_for` argument. It'll still work. (And probably it'll compile faster because this solution will require less template instantiations)

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: template argument gives an explicit "argument will be passed in compile-time", which is not always true for simply passing a constexpr argument into a constexpr function (as constexpr is not consteval). Imo current implementation is better for end users as it's easier to understand.

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: Proposed changes are also consistent with current static_for implementation that accepts a range as template arguments, not regular arguments.

File: base/base/constexpr_helpers.h (hunk 3 above)
Comment (novikd) [reply base/base/constexpr_helpers.h:10]: 1. If I want to iterate over all elements I don't want to check whether I need to continue processing or not. C++ ideology say "you don’t pay for what you don’t use".
2. I think it's possible to solve this problem by one simple `if constexpr` in `static_for`. It'll not make anything complex.
3. I think that anybody who uses `static_for` will expect it to iterate over all element if function returns void type. But you can add as a comment that if that person wants to be able to break the execution of the loop if may use another return type.

Comment (myrrc) [reply base/base/constexpr_helpers.h:10]: Ok, I guess you're right here. I'll change the solution to reflect void return types (but won't change any of the existing static_for calls), is it fine for you?

**Comment by novikd** [comment]
You can introduce an overload for `checkWithConstraints` instead of using `||` operator to solve your problem with:

https://github.com/ClickHouse/ClickHouse/pull/29690/files#diff-7e4a3be312e57980005947b14af5df111a4cf40ab2abfe9752b936ab7363a178R145

**Comment by myrrc** [comment]
I still think that a distinct CTArray type is more appropriate here and for further PRs.

**Review by novikd — CHANGES_REQUESTED** [review]
After the conversation I totally sure there is no any reason to pass array as template argument.

**Comment by myrrc** [comment]
This would possibly lead to iteration over array values in runtime, which is worse than proposed solution, I disagree.

**Comment by myrrc** [comment]
Also, this makes unpossible/inconsistend the overload that iterates over enums, which makes this PR useless.

File: base/base/CTArray.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: And why we can't make it `consteval`?

[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: Probably in that case users should not call `static_for` at all.

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: You say that static_for should be a more general solution and then say "users should not call ... at all", which does not look consistent. I've linked an example how other users can use static_for with template arguments to build higher abstractions (e.g. dispatchOverTypes) without copy-paste.

[the remark this answers is not in the document]
Comment (novikd) [reply base/base/CTArray.h:9]: `static_for` allows to perform loop in the **compile time**. If loop body (function passed to `static_for`) can't be evaluated in the compile time, it's completely OK to have compilation error, because it that case user should use `std::for_each`.

[the remark this answers is not in the document]
Comment (myrrc) [reply base/base/CTArray.h:9]: static_for is essentially a switch-case statement (as taken from D). It's not OK to have a CE when you can achieve switch-case semantics.

**Comment by myrrc** [comment]
Sadly, I don't have time to finish this PR, is someone is interested -- the branch will not be deleted.
